File: daemon/src/routes/request.ts
import type { Context } from 'hono'
import type { z } from 'zod'
import { AppError } from '@daemon/types/error'
import type { DocRoot } from '@daemon/services/Tree'
import type { DiffBasis } from '@daemon/utils/git'
import { ACTOR_HEADER, parseActor, type Actor } from '@daemon/types/ledger'
import { rootSchema } from './schemas'

export class BadRequest extends AppError {}

export async function parse<T>(c: Context, schema: z.ZodType<T>): Promise<T> {
  const raw = await c.req.json().catch(() => {
    throw new BadRequest('body must be JSON')
  })
  const result = schema.safeParse(raw)
  if (!result.success)
    throw new BadRequest(result.error.issues.map((i) => i.message).join('; '))
  return result.data
}

export function query(c: Context, name: string): string {
  const value = c.req.query(name)
  if (!value) throw new BadRequest(`missing ${name}`)
  return value
}

/** Which tree a GET is asking about. Every read names one, the same way every write does. */
export function root(c: Context): DocRoot {
  const result = rootSchema.safeParse(c.req.query('root'))
  if (!result.success) throw new BadRequest('root must be "project" or "repo:<name>"')
  return result.data
}

/** Who says they are doing this, for the ledger. The header is a claim and is read as one:
 *  an absent one is a person, which is what the editor's save is, and one that will not parse
 *  is nobody rather than a guess. */
export function actor(c: Context): Actor {
  return parseActor(c.req.header(ACTOR_HEADER))
}

/** Which two points a comparison is between. Unsaid is the branches as they stand, which is
 *  what the app opens on and what every caller before this one meant. */
export function basis(c: Context): DiffBasis {
  return c.req.query('basis') === 'split' ? 'split' : 'now'
}

File: daemon-go/internal/api/request.go
// What a handler reads out of a request. A body that is not JSON and a body that is JSON but
// not the shape asked for are two different complaints, and both are the caller's fault rather
// than a bug.

package api

import (
	"encoding/json"
	"io"
	"net/http"

	"github.com/broodmotherai/broodmother/daemon-go/internal/apperr"
	"github.com/broodmotherai/broodmother/daemon-go/internal/ledger"
)

// bodies larger than this are not a request anybody meant to send. The routes that take real
// content take it by path rather than in a body.
const maxBody = 8 << 20

func body(r *http.Request) (json.RawMessage, error) {
	raw, err := io.ReadAll(io.LimitReader(r.Body, maxBody))
	if err != nil {
		return nil, apperr.BadRequestf("body must be JSON")
	}
	var held json.RawMessage
	if json.Unmarshal(raw, &held) != nil {
		return nil, apperr.BadRequestf("body must be JSON")
	}
	return held, nil
}

// parse reads a body of the shape a route takes. `of` proves the shape and names what is wrong
// with it, the way each route's schema does in the TypeScript — in different words, since those
// are zod's and are read by a person rather than by a caller.
func parse[T any](r *http.Request, of func(json.RawMessage) (T, error)) (T, error) {
	var nothing T
	raw, err := body(r)
	if err != nil {
		return nothing, err
	}
	return of(raw)
}

// named is a `{ "<key>": "<a name>" }` body, which is most of what this API is asked with.
func named(key string) func(json.RawMessage) (string, error) {
	return func(raw json.RawMessage) (string, error) {
		var held map[string]json.RawMessage
		if json.Unmarshal(raw, &held) != nil || held == nil {
			return "", apperr.BadRequestf("body must be an object")
		}
		var value string
		if json.Unmarshal(orNull(held[key]), &value) != nil || value == "" {
			return "", apperr.BadRequestf("%s must be a name", key)
		}
		return value, nil
	}
}

func orNull(data json.RawMessage) json.RawMessage {
	if data == nil {
		return json.RawMessage("null")
	}
	return data
}

// query is a parameter a GET names in its URL.
func query(r *http.Request, name string) (string, error) {
	value := r.URL.Query().Get(name)
	if value == "" {
		return "", apperr.BadRequestf("missing %s", name)
	}
	return value, nil
}

// actorOf is who says they are doing this, for the ledger. The header is a claim and is read as
// one: an absent one is a person, which is what the editor's save is, and one that will not parse
// is nobody rather than a guess.
func actorOf(r *http.Request) ledger.Actor {
	return ledger.ParseActor(r.Header.Get(ledger.Header))
}
